lib/middleware/validateSession.js

```diff
@@ -0,0 +1,70 @@
+var getSessionVal = function(obj, key) {
+  key = key.split('.');
+
+  for (var i = 0; i < key.length; i++) {
+    var curKey = key[i];
+
+    if (curKey in obj) {
+      obj = obj[curKey];
+    } else {
+      return null;
+    }
+  }
+
+  return obj;
+};
+
+module.exports = function validateSession(mediator, mbaasApi, opts) {
+
+  return function validateSessionMiddleware(req, res, next) {
+    var session,
+      sessionToken;
+
+    opts = {
+      sessionKey: opts.sessionKey || "fh_params",
+      sessionIdKey: opts.sessionIdKey || "__fh.sessiontoken",
+      sessionHeaders: opts.sessionHeaders || "x-fh-sessiontoken"
+    };
+
+    session = getSessionVal(req, opts.sessionKey) || {};
+    sessionToken = getSessionVal(session, opts.sessionIdKey) || req.headers[opts.sessionHeaders] || null;
+
+    if (!sessionToken) {
+      console.log("No session token available");
+      return res.status(401).json(new Error("Unauthorized"));
+    }
+
+    //Check cache for session
+    mbaasApi.session.get(sessionToken, function(err, sessionData) {
+      if (err) {
+        console.log(err);
+        return res.status(500).json(err);
+      }
+
+      if (sessionData) {
+        return next();
+      }
+
+      mediator.request("wfm:user:session:validate", sessionToken)
```

Comment:
This function can be split in two.
e.g. 

You could implement two different functions and abstract them from this larger function

`checkSessionCache` and `verifySession`

`checkSession` - would check the session cache entry. If it exists, then the middleware can move on

`verfiySession` - Will publish the topic to the mediator and set the cache value if valid.

These functions can then be implemented in different files. It would make it easier to unit test and read.